### ISSUE TITLE
Reflection: cannot programmatically give refl

### ISSUE THREAD

**Issue by omelkonian** [agda/agda#5738]
Consider the simple no-op macro that just unifies the hole with a given term:
```agda
macro
  give : Term → Hole → TC ⊤
  give = flip unify
```
We would expect this to work the same as directly writing the argument term, but this is not true even for a trivial equality by reflexivity:
```agda
_ : 5 ≡ 5
_ = give refl
```
```bash
# Yellow warnings
_a_269 : Level
_A_270 : Set _a_269
_x_271 : _A_270
```
If we examine the argument term and its type we discover that the problem lies with unresolved metavariables in `refl`'s implicit arguments and inferred type:
```agda
give t hole = do
  goal ← inferType hole
  ty ← inferType t
  print $ "Goal: " ◇ show goal ◇ "\nTerm: " ◇ show t ◇ "\nType: " ◇ show ty
  unify hole t
```
```bash
Goal: _≡_ {lzero} {Nat} 5 5
Term: refl {unknown} {unknown} {unknown}
Type: _≡_ {_492} {_493}  _494 _494
```
Just to verify, if we annotate `refl`'s type before passing it as an argument everything's fine, while annotating outside the macro call is futile:
```agda
_ = give (5 ≡ 5 ∋ refl) -- succeeds
_ = 5 ≡ 5 ∋ give refl   -- fails
```
Is this a bug in `unify : Term → Term → TC ⊤`?
I would expect after the call to `unify` variables `_492/_493/_494` are resolved and replaced by `lzero/Nat/5`. 

However, we get the expected behaviour on the types, i.e. f we unify the goal with the inferred type and inspect the updated term, we get the expected resolved type:
```agda
give t hole = do
  goal ← inferType hole
  unify ty goal
  ty ← reduce ty
  print $ "Goal: " ◇ show goal ◇ "\nTerm: " ◇ show t ◇ "\nType: " ◇ show ty
  unify hole t
```
```bash
Goal: _≡_ {lzero} {Nat} 5 5
Term: refl {unknown} {unknown} {unknown}
Type: _≡_ {lzero} {Nat}  5 5
```   
Why doesn't the same apply for the argument term?

*Agda: 2.6.2.1*

**Comment by jespercockx** [comment]
This is a recurring issue / design flaw. What's going on is that Agda will first typecheck the term `refl` with a fresh meta as its type, then quote the result, and finally execute the call to `unify`. However, in the quoting process any link with the original term is lost, so there is no connection between the `refl` that you wrote in the code and the `refl` that ends up in the elaborated term. It kind of makes sense if you know how macros are desugared to calls to `quoteTerm` and `unquoteTerm`, but it does trip up a lot of newcomers.

The easiest way to fix your macro that I can think of is by doing the quoting manually and adding a type annotation to the quoted term:
```agda
open import Data.Unit
open import Data.Nat.Base
open import Function using (_∋_)
open import Reflection
open import Reflection.TypeChecking.Monad.Syntax
open import Relation.Binary.PropositionalEquality

macro
  give : ∀ {a} {A : Set a} → A → Term → TC ⊤
  give {a} {A} x hole = do
    `x ← quoteTC (A ∋ x)
    unify hole `x

_ : 5 ≡ 5
_ = give refl
```
This allows Agda to learn that the type of the argument of `give` is the same as the type of the goal.

In general it feels to me that there should be a better way to design a reflection API that separates the notions of "having a handle to a term that allows you to pass it to operations such as `unify`" from "having an explicit syntactic representation of that term as an element of type `Term`". It could be interesting to experiment with adding a type for representing the former thing in Agda, since we currently only have very indirect ways to work with it.

**Comment by JacquesCarette** [comment]
That last paragraph is extremely interesting indeed. What comes to mind is an abstract interface for each representation (handle on term, syntax representation). The internal representations could be extremely similar, but still offer a quite different external API of functionality.

I think of it like the difference between a directed graph and a Quiver. Formally, they are identical objects, but intentionally (yes, with a 't'), they are quite different, because you use them in different ways, for different means. The choice of name communicates a whole lot of context that the object itself does not.  In other words, there could be more than one type where right now there is just `Term`.

**Comment by omelkonian** [comment]
> However, in the quoting process any link with the original term is lost, so there is no connection between the `refl` that you wrote in the code and the `refl` that ends up in the elaborated term.

This still bugs me, as I would expect my macro to work even without the connection you mention.
We have a hole of type `5 ≡ 5` and a term given by the user of type `refl {unknown} {unknown} {unknown}`:
shouldn't `unify` itself replace the `unknown`s inside `refl` appropriately, without the need for your quoting workaround?

**Comment by jespercockx** [comment]
Yes, those metavariables are indeed solved. The problem is just that *before* running the macro, Agda has already typechecked `refl` at an unknown type, and the metas created during this process have not been solved. So the problem is really that Agda always typechecks arguments (at unknown type) before passing them to a macro.

**Comment by omelkonian** [comment]
Aha, now I think I get it! I had confused the notion of a term at unknown type (`unknown`) with a meta-variable (`meta ?152`).

So, ideally we would like the term to be `refl {meta_?178} {meta_?179} {meta_?180}` inside the macro, right?

**Comment by tomsmeding** [comment]
@jespercockx Thank you for your super helpful example above in https://github.com/agda/agda/issues/5738#issuecomment-1013295556. I was trying to write a macro that, given a proof of `a ≡ b` and a goal that mentions `a`, allows me to continue with the goal where all occurrences of `a` are replaced with `b`. This is kind of what the built-in `rewrite` syntax does, but I would like this halfway down a function, not only at the top level.

I was stuck because it seemed impossible to pass a hole as argument to a macro (and have the macro return that hole as part of its result), even if the macro knows exactly what the type of the "continuation" must be. However, by passing the continuation as a value, not a Term, and unifying its (implicitly passed in) type with the computed type inside the macro, it works perfectly.

This experience would be significantly improved with better support for what was being discussed in this issue (e.g. @omelkonian's suggestion of explicitly passing in metavariables to a macro so that the returned term can contain the _same_ metavariables). But at least there is a workaround for my use case.

(For interested readers: a very rough initial version of the macro I described is [here](https://paste.tomsmeding.com/Y2DgF20M). I may clean it up later; feel free to send me a message to check if I have.)
